NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -68,44 +148,42 @@ function CategoryProduct(pairs::Pair...)
   return CategoryProduct(NamedTuple{keys}(vals))
 end
 
-function categories_fusion_rule(A::NamedTuple, B::NamedTuple)
-  qs = [A]
-  for (la, lb) in zip(pairs(intersect_keys(A, B)), pairs(intersect_keys(B, A)))
-    @assert la[1] == lb[1]
-    fused_vals = ⊗(la[2], lb[2])
-    qs = [union_keys((; la[1] => v), q) for v in fused_vals for q in qs]
-  end
-  # Include sectors of B not in A
-  qs = [union_keys(q, B) for q in qs]
-  return qs
-end
-
 function categories_equal(A::NamedTuple, B::NamedTuple)
   common_categories = zip(pairs(intersect_keys(A, B)), pairs(intersect_keys(B, A)))
   common_categories_match = all(nl -> (nl[1] == nl[2]), common_categories)
   unique_categories_zero = all(l -> istrivial(l), symdiff_keys(A, B))
   return common_categories_match && unique_categories_zero
 end
 
-#
-# Ordered implementation
-#
-
-CategoryProduct(t::Tuple) = _CategoryProduct(t)
-CategoryProduct(cats::AbstractCategory...) = CategoryProduct((cats...,))
+function trivial(::Type{<:CategoryProduct{NT}}) where {Keys,NT<:NamedTuple{Keys}}
+  return reduce(×, (ntuple(i -> (; Keys[i] => trivial(fieldtype(NT, i))), fieldcount(NT))))
+end
+trivial(::Type{<:CategoryProduct{<:NamedTuple{()}}}) = sector()  # Empty NamedTuple
+
+# allow ⊗ for different types in NamedTuple
+function categories_fusion_rule(cats1::NamedTuple, cats2::NamedTuple)
+  diff_cat = CategoryProduct(symdiff_keys(cats1, cats2))
+  nt1 = intersect_keys(cats1, cats2)
+  shared1 = ntuple(i -> (; keys(nt1)[i] => values(nt1)[i]), length(nt1))
+  nt2 = intersect_keys(cats2, cats1)
+  shared2 = ntuple(i -> (; keys(nt2)[i] => values(nt2)[i]), length(nt2))
+  return diff_cat × categories_fusion_rule(shared1, shared2)
+end
 
-categories_equal(o1::Tuple, o2::Tuple) = (o1 == o2)
+# abelian fusion of one category
+function fusion_rule(::AbelianGroup, cats1::NT, cats2::NT) where {NT<:NamedTuple}
+  fused = fusion_rule(only(values(cats1)), only(values(cats2)))
+  return sector(only(keys(cats1)) => fused)
+end
 
-function categories_fusion_rule(o1::Tuple, o2::Tuple)
-  N = length(o1)
-  length(o2) == N ||
-    throw(DimensionMismatch("Ordered CategoryProduct must have same size in ⊗"))
-  os = [o1]
-  replace(o, n, val) = ntuple(m -> (m == n) ? val : o[m], length(o))
-  for n in 1:N
-    os = [replace(o, n, f) for f in ⊗(o1[n], o2[n]) for o in os]
+# generic fusion of one category
+function fusion_rule(::SymmetryStyle, cats1::NT, cats2::NT) where {NT<:NamedTuple}
+  fused = fusion_rule(only(values(cats1)), only(values(cats2)))
+  key = only(keys(cats1))
+  v = Vector{Pair{CategoryProduct{NT},Int64}}()
+  for la in BlockArrays.blocklengths(fused)
+    push!(v, sector(key => LabelledNumbers.label(la)) => LabelledNumbers.unlabel(la))
   end
-  return os
+  g = GradedAxes.gradedrange(v)
+  return g
 end
```

Comment:
I'm having trouble understanding why this code is designed this way, it seems like three functions that are pretty similar, maybe this can be simplified?

Reply:
As discussed, the issue is for `NonAbelianGroup` or `NonGroupCategory`, `fusion_rule` produces a `GradedAxis`. Then one needs to construct a new `GradedAxis` and substitute labels with `NamedTuple`. I only managed to do this for a single category, so I first split a `CategoryProduct` into tuples of of one category. I am not happy with the current design but I struggle to improve it.

Reply:
Finally found a solution :)